rfcs/53-generalized-hierarchy.md

```diff
@@ -0,0 +1,273 @@
+  prev: Option<Entity>,
+  next: Option<Entity>
+  first_child: Option<Entity>,
+}
+```
+
+This type has no public constructors and has no public APIs that allow for
+mutation of any of the internal data, only getters. The consistency between
+multiple instances of the component is maintained by a set of dedicated hierachy
+management commands. The following should be added as extensions to
+EntityCommands:
+
+  - `set_parent(Option<Entity>)` - sets the parent of a given entity.
+      - If `Hierarchy` is on neither parent or child, it will be added and both
+        will be atomically updated to reflect this change.
+      - If the parent already had children, the new child will be prepended at
+        the start of the linked list, and the sibling pointers of both will be
+        updated.
+      - If the child already had a parent, that previous parent will be updated
+        too.
+      - A HierarchyEvent is added to the queue to reflect this update.
+  - `move_to_first(Entity)` - Moves one of the children to the front.
+  - `move_to_last(Entity)` - Moves one of the children to the end.
+  - `remove_all_children()` - Disconnects all children from a parent. All
+    children become roots.
+      - A HierarchyEvent is added to the queue to reflect this update.
+
+All of these commands will also update the four marker components appropriately.
+Optionally, we could also mark these components as changed (even though they're
+ZSTs) to leverage change detection.
+
+One optional extension to this design is to create dedicated `HierarchyQuery<Q,
+F>`, which is a custom SystemParam that wraps multiple Query objects to simplify
+querying for items in the hierarchy:
+
+ - `iter_descendants(_mut)` - returns an iterator over `(parent, child)` results
+   from a query parameter. This does a depth-first traversal over the hierarchy
+   starting at some entity.
+ - `iter_ancestors(_mut)` - returns an iterator over `(parent, child)` results
+   from a query parameter. This bubbles up queries from the bottom to the top
+   starting at some entity.
+
+As these operations are handled at any command buffer flush. The hierarchy
+maintenance system can be removed.
+
+## Benefits
+The biggest benefit to this new design is guarenteed global consistency. The
+hierarchy cannot be arbitrarily mutated between synchronization points.
+
+The second benefit is that any update to the hierarchy is "atomic" and is `O(1)`
+for the majority of the operations handled.
+
+## Drawbacks
+There are a notable number of drawbacks to this design:
+
+ - Counting children is now `O(n)`. Finding and editing children order is still
+   `O(n)`, that has not changed.
+ - The fact that the hierarchy is not optimally stored in memory is still an
+   issue, and all hiearchy traversals require heavy random access into memory.
+ - Updates are still not immediately visible and deferred until the next command
+   buffer flushes.
+ - Hierarchy updates are now single threaded.
```

Comment:
Can't we can add internal parallelism to resolve this?

Reply:
Unless we somehow parallelize commands, I don't see how this is possible.

Reply:
With the advent of https://github.com/bevyengine/bevy/pull/4749, this is mostly (?) mitigated. The exclusive nature of commands is likely not going to change anytime soon.